src/lcals-kokkos/HYDRO_2D-Kokkos.cpp

```diff
@@ -0,0 +1,129 @@
+  auto zadat_view = getViewFromPointer(zadat, kn, jn);
+  auto zbdat_view = getViewFromPointer(zbdat, kn, jn);
+  auto zmdat_view = getViewFromPointer(zmdat, kn, jn);
+  auto zpdat_view = getViewFromPointer(zpdat, kn, jn);
+  auto zqdat_view = getViewFromPointer(zqdat, kn, jn);
+  auto zrdat_view = getViewFromPointer(zrdat, kn, jn);
+  auto zudat_view = getViewFromPointer(zudat, kn, jn);
+  auto zvdat_view = getViewFromPointer(zvdat, kn, jn);
+  auto zzdat_view = getViewFromPointer(zzdat, kn, jn);
+
+  // Wrap output pointers into Kokkos::Views
+  auto zroutdat_view = getViewFromPointer(zroutdat, kn, jn);
+  auto zzoutdat_view = getViewFromPointer(zzoutdat, kn, jn);
+
+  switch (vid) {
+
+  case Kokkos_Lambda: {
+
+    Kokkos::fence();
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      // Use MDRangePolicy for multidimensional arrays
+
+      Kokkos::parallel_for(
+          "HYDRO_2D_Kokkos Kokkos_Lambda--BODY1",
+          Kokkos::MDRangePolicy<Kokkos::Rank<2>>({kbeg, jbeg}, {kend, jend}),
+          KOKKOS_LAMBDA(int64_t k, int64_t j) {
+            zadat_view(k, j) =
+                (zpdat_view(k + 1, j - 1) + zqdat_view(k + 1, j - 1) -
+                 zpdat_view(k, j - 1) - zqdat_view(k, j - 1)) *
+                (zrdat_view(k, j) + zrdat_view(k, j - 1)) /
+                (zmdat_view(k, j - 1) + zmdat_view(k + 1, j - 1));
+
+            zbdat_view(k, j) = (zpdat_view(k, j - 1) + zqdat_view(k, j - 1) -
+                                zpdat_view(k, j) - zqdat_view(k, j)) *
+                               (zrdat_view(k, j) + zrdat_view(k - 1, j)) /
+                               (zmdat_view(k, j) + zmdat_view(k, j - 1));
+          });
+
+      Kokkos::parallel_for(
+          "HYDRO_2D_Kokkos Kokkos_Lambda--BODY2",
+          Kokkos::MDRangePolicy<Kokkos::Rank<2>>({kbeg, jbeg}, {kend, jend}),
+          KOKKOS_LAMBDA(int64_t k, int64_t j) {
+            zudat_view(k, j) +=
+                s *
+                (zadat_view(k, j) * (zzdat_view(k, j) - zzdat_view(k, j + 1)) -
+                 zadat_view(k, j - 1) *
+                     (zzdat_view(k, j) - zzdat_view(k, j - 1)) -
+                 zbdat_view(k, j) * (zzdat_view(k, j) - zzdat_view(k - 1, j)) +
+                 zbdat_view(k + 1, j) *
+                     (zzdat_view(k, j) - zzdat_view(k + 1, j)));
+            zvdat_view(k, j) +=
+                s *
+                (zadat_view(k, j) * (zrdat_view(k, j) - zrdat_view(k, j + 1)) -
+                 zadat_view(k, j - 1) *
+                     (zrdat_view(k, j) - zrdat_view(k, j - 1)) -
+                 zbdat_view(k, j) * (zrdat_view(k, j) - zrdat_view(k - 1, j)) +
+                 zbdat_view(k + 1, j) *
+                     (zrdat_view(k, j) - zrdat_view(k + 1, j)));
+          });
+
+      Kokkos::parallel_for(
+          "HYDRO_2D_Kokkos Kokkos_Lambda--BODY3",
+          Kokkos::MDRangePolicy<Kokkos::Rank<2>>({kbeg, jbeg}, {kend, jend}),
```

Comment:
I like the compactness of the policy. But curious, hows does the thread mapping work here for the device? @ajpowelsnl  is this global thread indexing?

Reply:
The mapping is architecture - specific and not guaranteed by Kokkos.  However, there are ordering template parameters, such as Kokkos::View Layouts, which are defaulted.  In practice , `MDRange` does tiling and each thread is assigned to a CPU thread or a GPU Cuda bloc, or the equivalent (in other backends). On GPU, consecutive iterations as defined by iteration order are assigned to consecutive threads in the CUDA bloc.